subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java

```diff
@@ -138,47 +142,67 @@ private void updateWatchedHierarchies(SnapshotHierarchy root) {
+                stopWatchingProbeForHierarchy(probedRoot);
+                probeRegistry.disarmWatchProbe(probedRoot);
+            });
+
         if (!hierarchiesToStopWatching.isEmpty()) {
-            hierarchiesToStopWatching.forEach(probeRegistry::disarmWatchProbe);
             stopWatchingHierarchies(hierarchiesToStopWatching);
         }
         if (!hierarchiesToStartWatching.isEmpty()) {
             hierarchiesToStartWatching.forEach(locationToWatchValidator::validateLocationToWatch);
             startWatchingHierarchies(hierarchiesToStartWatching);
-            hierarchiesToStartWatching.forEach(probeRegistry::armWatchProbe);
         }
+
+        probedRoots.stream()
+            .filter(newProbedRoot -> !oldProbedRoots.contains(newProbedRoot))
+            .forEach(probedRoot -> {
+                startWatchingProbeForHierarchy(probedRoot);
+                probeRegistry.armWatchProbe(probedRoot);
+            });
+
         LOGGER.info("Watching {} directory hierarchies to track changes", watchedRoots.size());
     }
 
     protected abstract void startWatchingHierarchies(Collection<File> hierarchiesToWatch);
+    protected abstract void startWatchingProbeForHierarchy(File hierarchyToWatch);
     protected abstract void stopWatchingHierarchies(Collection<File> hierarchiesToWatch);
+    protected abstract void stopWatchingProbeForHierarchy(File hierarchyToWatch);
 
     public interface FileSystemLocationToWatchValidator {
         FileSystemLocationToWatchValidator NO_VALIDATION = location -> {};
 
         void validateLocationToWatch(File location);
     }
 
+    /**
+     * Resolves the watched {@link FileHierarchySet} and collects the roots of each hierarchy
+     * that needs to be watched.
+     *
+     * We need both because with nested hierarchies the outermost hierarchy will "hide" all
+     * inner hierarchies in the {@link FileHierarchySet}, yet we will still need to prove
+     * even the inner hierarchies.
+     */
     @VisibleForTesting
-    static FileHierarchySet resolveWatchedHierarchies(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot) {
+    static FileHierarchySet resolveWatchedHierarchies(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot, ImmutableSet.Builder<File> rootsToProbe) {
         FileHierarchySet watchedHierarchies = DefaultFileHierarchySet.of();
         for (File watchableHierarchy : watchableHierarchies.getRecentlyUsedHierarchies()) {
             String watchableHierarchyPath = watchableHierarchy.getAbsolutePath();
-            if (watchedHierarchies.contains(watchableHierarchyPath)) {
-                continue;
-            }
-
-            if (hasNoContent(vfsRoot.rootSnapshotsUnder(watchableHierarchyPath), watchableHierarchies, watchedHierarchies)) {
+            if (hasNoContent(vfsRoot.rootSnapshotsUnder(watchableHierarchyPath), watchableHierarchies)) {
                 continue;
             }
+            rootsToProbe.add(watchableHierarchy);
```

Comment:
If we do it like you propose here, then we would not probe an inner watchable hierarchy of a watched root if it doesn't contain anything to watch. I wonder if that is good or not.

I think it may be simpler to do a
```
rootsToProbe = watchableHierarchies.stream()
  .filter(watchedHierarchies::contains)
  .collect(toSet());
```

after determining the hierarchies to watch. WDYT?

Reply:
> If we do it like you propose here, then we would not probe an inner watchable hierarchy of a watched root if it doesn't contain anything to watch. I wonder if that is good or not.

FTR, we just discussed that we'd be fine with this.

I like the simplification you propose, let's do it.

Reply:
I am confused now, the simplification means that we would probe an inner watchable hierarchy if it doesn't contain anything to watch.